### ISSUE TITLE
what is the format of a hash?

### ISSUE THREAD

**Issue by dominictarr** [pijul/pijul#6]
The thing I really want to know, to understand how pijul works,
is how other patches refer to each other, what sort of "data structure" is pijul?
A good way to do that is to describe the format.

**Comment by pijul** [comment]
In file [pijul/libpijul/src/patch.rs](pijul/libpijul/src/patch.rs)

**Comment by dominictarr** [comment]
So, a partially ordered set is called a _poset_. They sound quite interesting - I don't really know much about them - but they do sound interesting - a trees, sets, and lists can all be represented as posets, where lists and sets are special cases (totally ordered and unordered respectively).

I was given this document about posets [pod.pdf](https://github.com/pijul/pijul/files/192693/pod.pdf)

Hmm, so to represent the state of a repo, I hash the set of patches in that repo - although, you could since hashes already point to hashes they depend on, you'd only need to refer to the _highest_ (poset context) hashes, the tips of the tree. Which seems like it would be somewhat smaller than the full list of patches. it also occurs to me that you could regard the patch set as a "document" that is also being edited, and patch the patchset, applying all the same logic.

**Comment by pijul** [comment]
>  although, you could since hashes already point to hashes they depend on, you'd only need to refer to the highest (poset context) hashes, the tips of the tree

We've not implemented this yet, but it's in the plans, to improve the complexity of push and pull.

>  you could regard the patch set as a "document" that is also being edited, and patch the patchset, applying all the same logic

Sure, the problem, as always, is conflicts: what if Alice merges patch set A to B, while Bob pushes to branch A?

**Comment by dominictarr** [comment]
hmm, well it's much easier to patch/merge sets than lists, because the order doesn't matter or (as, in this cas is explicit in the set's items), so you'll only get a conflict if a patch is removed from one and added to another, right?

How does a patch that resolves a conflict work? is it different than a normal patch in any way?

Is the pijul website/docs in a repo? I'd like to make a pull request to improve the documentation once I am satisfied i understand how it works.

**Comment by pijul** [comment]
We have a darcs repository, but it's terribly messy at the moment. Thanks for the suggestion, we'll open it when it's clean.

**Comment by dominictarr** [comment]
I found http://pijul.org/2016/03/03/functional-database.html but I wish it actually described what it is. I'm guessing it's basically a more efficient LMDB. you write _about_ but you don't describe _what it does_ (like, what properties or api it has)

**Comment by dominictarr** [comment]
ah, okay so what is the algorithm to see whether there is a conflict?

**Comment by P-E-Meunier** [comment]
> I found http://pijul.org/2016/03/03/functional-database.html but I wish it actually described what it is. I'm guessing it's basically a more efficient LMDB. you write about but you don't describe what it does (like, what properties or api it has) 

Right, there was no API documentation at the time I wrote that post. Sanakirja is not really meant to be more efficient than LMDB, it's just persistent, meaning it implements the same semantics as e.g. Haskell data structures: updating it does not destroy the previous version.
LMDB has that property, to some extent (although they use B+ trees, which cannot really be persistent), but they don't expose that API to the user, essentially because the pages are not reference-counted (or otherwise garbage-collected).

Now Sanakirja has an official page: [http://pijul.org/sanakirja](http://pijul.org/sanakirja)

**Comment by P-E-Meunier** [comment]
Here you go for the repository of pijul.org:
http://hub.darcs.net/pmeunier/pijul.org/

Right, not on github, but the following should work
- Go to [http://hub.darcs.net/pmeunier/pijul.org/fork](http://hub.darcs.net/pmeunier/pijul.org/fork)
- Click "Fork", and follow the instructions.
- ```darcs get http://hub.darcs.net/you/pijul.org```
- edit files, possibly ```darcs add```ing new files.
- ```darcs record```
- ```darcs push```

Alternatively, you can also write your username in a reply here, and I'll add you as an author, so that you can push directly there.
